### PR TITLE
Update episode 08-development-setup

### LINKED ISSUE #167: Source installation of ESMValTool from GitHub.
Episode to install ESMValTool from the GItHub source.

### PR DESCRIPTION
**Pull request by SarahAlidoost**
# Pull Request checklist

We appreciate your time and effort to improve the tutorial. Please keep in mind that lesson maintainers are volunteers and it may be some time before they can respond to your contribution.

* * *

## Before you start

- [x] Read [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md).
- [x] Create an [issue](https://github.com/ESMValGroup/tutorial/issues) to discuss your idea. This allows your contributions to be incorporated into the tutorial.

## Tasks

- [x] Give this pull request a descriptive title.
- [x] If you are contributing to existing lesson materials, please make sure the content conforms to the `Lesson development` section in [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md) and does not contain any spelling or grammatical errors.
- [x] If you are making a new episode, please make sure the content conforms to the `Lesson organization` and `Lesson formatting` sections in [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md) and does not contain any spelling or grammatical errors.
- [x] Preferably Codacy checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why.
- [x] Preview changes on your machine before pushing them to GitHub by running `make serve`, alternatively `make docker-serve`. Please see the `Previewing your changes locally` section in [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md) for installation instructions.
- [x] All code instructions have been tested.

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #167 #164
Follow up issue #190

### CODE REVIEW

**Comment by SarahAlidoost** [comment]
The remaining CI errors are related to the links at the end of the file and the `markdown` code blocks. I couldn't find a better-supported language by the template, see the list [here](https://github.com/ESMValGroup/ESMValTool_Tutorial/blob/ff50d10915459a51336bfdc94716b46cffbdbd4a/bin/lesson_check.py#L73).

**Review by Peter9192 — CHANGES_REQUESTED** [review]
Hey @SarahAlidoost, nice work! It's become much more extensive than I initially expected, but I think many people are gonna benefit from it. Since it covers quite a lot of topics, I have quite a few comments... but I tried to provide suggestions whereever I could :-)

File: _episodes/08-development-setup.md [hunk 1]
```diff
@@ -1,138 +1,394 @@
...
-By default, this command will create a folder called ESMValTool containing the
-source code of the tool.
+This command will ask your GitHub username and a personal **token** as password.
+Please follow instructions on
+[GitHub token authentication requirements][token-authentication-requirements]
+to create a personal access token.
+After the authentication, a folder called ``ESMValTool`` is created in your working directory.
+This folder contains the source code of the tool.
+To continue the installation, we move into the ``ESMValTool`` directory
+and check out the ``master`` branch:
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply _episodes/08-development-setup.md]: correct, it is replaced by a note that it is the default.

File: _episodes/08-development-setup.md [hunk 2]
```diff
@@ -1,138 +1,394 @@
...
->
-> Make sure that the master branch is checked out before continuing intallation:
-> ~~~bash
-> git checkout master
-> ~~~
-{: .callout}
+~~~bash
+cd ESMValTool
+git checkout master
+~~~
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply _episodes/08-development-setup.md:85]: I added the output of git clone and git status. Please let me know if it is not what you meant.

File: _episodes/08-development-setup.md [hunk 3]
```diff
@@ -1,138 +1,394 @@
...
-From now on, we will assume that the installation is going to be done through
-conda.
+> ## environment name
+>
+> If an ``esmvaltool`` environment is already created following the lesson
+> [Installation]({{ page.root }}{% link _episodes/02-installation.md %}),
+> first delete it or choose another name for the new environment in this lesson.
+> For more information see
+> [conda managing environments][manage-environments].
+{: .callout}
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply _episodes/08-development-setup.md]: Agree, I added your suggestion.

File: _episodes/08-development-setup.md [hunk 4]
```diff
@@ -1,138 +1,394 @@
...
 ~~~bash
-conda activate esmvaltool
+esmvaltool --help
 ~~~
 
-If you run into trouble, please try recreating the environment. More information
-can be found in the [conda documentation](https://docs.conda.io/en/latest/).
+If installation is successful, ESMValTool prints a help message to the console.
+For more details about development installation, see ESMValTool documentation on
+[install from source][install-from-source].
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply _episodes/08-development-setup.md:148]: Agree, I remove the link to the documentation from here.

File: _episodes/08-development-setup.md [hunk 5]
```diff
@@ -1,138 +1,394 @@
...
+- a new or updated recipe script, see lesson on
+[Writing your own recipe]({{ page.root }}{% link _episodes/05-preprocessor.md %})
+- a new or updated diagnostics script, see lesson on
+[Writing your own diagnostic script]
+- a new or updated cmorizer script, see lesson on
+[CMORization: Using observational datasets]
+- helping  with reviewing process of pull requests, see ESMValTool documentation on
+[Review of pull requests](https://docs.esmvaltool.org/en/latest/community/review.html)
+
+The next sections will explore the ways we can achieve this.
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply _episodes/08-development-setup.md]: Agree, the paragraph is moved to the top. I also added a paragraph similar to what you suggested.

File: _episodes/08-development-setup.md [hunk 6]
```diff
@@ -1,138 +1,394 @@
...
+**[issue](https://github.com/ESMValGroup/ESMValTool/issues)** in ESMValTool repository.
+Then, we create a new ``branch`` locally and start developing new codes.
+Once our development is finished, we can initiate a ``pull request``.
+For a full description of the GitHub workflow, please see ESMValTool documentation on
+[GitHub Workflow](https://docs.esmvaltool.org/en/latest/community/repository.html#github-workflow).
+
+The pull request will be tested, discussed and merged. This is called "**review process**".
+The process will take some effort and time to learn.
+However, a few “tools” i.e. command lines gets you a long way,
+and we’ll cover those essentials in the next sections.
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply _episodes/08-development-setup.md:197]: Nice suggestion! I added a tip similar to what you said here.

File: _episodes/08-development-setup.md [hunk 7]
```diff
@@ -1,138 +1,394 @@
...
+
+### Check code quality
+
+We aim to adhere to best practices and coding standards. The good news is that there are
+several tools that we can use to check our code against those standards.
+
+As an example, ``Pre-commit`` is a tool that checks your code for any invalid syntax
+and formatting errors. It also fixes some of those errors.
+To explore other tools, have a look at ESMValTool documentation on
+[Code style](https://docs.esmvaltool.org/en/latest/community/introduction.html#code-style).
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply _episodes/08-development-setup.md:232]: I added the list of the tools as you suggested. However, I didn't remove the link to the documentation as there is some info in the documentation that is beyond the scope of this lesson. Also, I think neither documentation nor this tutorial should cover details about these tools. Because they are related to best practices and code refinery lessons. We might just point users to those resources. This is also in line with what you mentioned about duplicate information.

File: _episodes/08-development-setup.md [hunk 8]
```diff
@@ -1,138 +1,394 @@
...
+
+Previous section introduced some tools to check code style and quality.
+What it hasen’t done is show us how to tell whether our code is getting the right answer.
+To achieve that, we can run tests using ``pytest`` locally:
 
 ~~~bash
-pip install --editable '.[develop]'
+conda activate esmvaltool
+cd ESMValTool
+git checkout your_branch_name
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply _episodes/08-development-setup.md]: right, they are removed.

File: _episodes/08-development-setup.md [hunk 9]
```diff
@@ -1,138 +1,394 @@
...
+>> We should check our codes regarding the usage of ``nx``.
+>> For now, let's assume that it is added by mistake and remove it.
+> {: .solution}
+{: .challenge}
+
+### Run unit tests
+
+Previous section introduced some tools to check code style and quality.
+What it hasen’t done is show us how to tell whether our code is getting the right answer.
+To achieve that, we can run tests using ``pytest`` locally:
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply _episodes/08-development-setup.md]: It is added as you suggested.

File: _episodes/08-development-setup.md [hunk 10]
```diff
@@ -1,138 +1,394 @@
...
+>>
+>> ~~~ bash
+>>
+>> E           esmvalcore._task.DiagnosticError: Cannot execute script
+>> '~/esmvaltool_tutorial/warming_stripes.py' (~/esmvaltool_tutorial/warming_stripes.py):
+>> file does not exist.
+>>~~~
+>>
+>> It seems that we need to set a correct path for the diagnostic script
+>> ``warming_stripes.py`` in our recipe!
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply _episodes/08-development-setup.md]: To address this comment, I added a background section in the contribution, and revise the text and exercises. Please have a look again and let me know if it is still confusing.

File: _episodes/08-development-setup.md [hunk 11]
```diff
@@ -1,138 +1,394 @@
...
+> {: .solution}
+{: .challenge}
 
-## Test the installation
+### Build documentation
 
-The next step is to check that the installation works properly.
-To do this, run the tool with:
+Documentations are available on
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply _episodes/08-development-setup.md:371]: An introductory sentence is now added.

File: _episodes/08-development-setup.md [hunk 12]
```diff
@@ -1,138 +1,394 @@
...
+[docs.esmvaltool.org](https://docs.esmvaltool.org/en/latest/index.html).
+The source files are located in ``ESMValTool/doc/sphinx/source/``.
+To build documentations locally, we run:
 
 ~~~bash
-esmvaltool --help
+conda activate esmvaltool
+cd ESMValTool
+git checkout your_branch_name
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply _episodes/08-development-setup.md]: Absolutely right. It is now removed.

File: _episodes/08-development-setup.md [hunk 13]
```diff
@@ -1,138 +1,394 @@
...
+> Now, we create a documentation file ``recipe_warming_stripes.rst`` for this recipe:
+>
+> ~~~bash
+> nano doc/sphinx/source/recipes/recipe_warming_stripes.rst
+> ~~~
+>
+> Add a reference i.e. ``.. _recipe_warming_stripes:``, a section title
+> and some text about the recipe like:
+>
+> ~~~markdown
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply _episodes/08-development-setup.md]: I think `markdown` is one of the languages supported for the code block. However, our template does not support too many languages see [here](https://github.com/ESMValGroup/ESMValTool_Tutorial/blob/ff50d10915459a51336bfdc94716b46cffbdbd4a/bin/lesson_check.py#L73) . I replaced `~~~ `as you suggested. The test is passed but it does not render nicely as before.

**Comment by SarahAlidoost** [comment]
> Hey @SarahAlidoost, nice work! It's become much more extensive than I initially expected, but I think many people are gonna benefit from it. Since it covers quite a lot of topics, I have quite a few comments... but I tried to provide suggestions whereever I could :-)

Thank you for reviewing and helpful suggestions. I addressed your comments. Please have another look.

File: _episodes/08-development-setup.md [hunk 14]
```diff
@@ -1,138 +1,461 @@
...
+For more details about development installation, see ESMValTool documentation on
+[install from source][install-from-source].
+
+> ## Checking the development installation
+>
+> We can use the command ``conda list`` to list installed packages in
+> the ``esmvaltool`` environment.
+> Use this command to check that ESMValTool is installed in a ``develop`` mode.
+>
+> **Tips**: see the
```
Comment (Peter9192) [inline _episodes/08-development-setup.md:156]:
```suggestion
> **Tip**: see the
```

File: _episodes/08-development-setup.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (Peter9192) [reply _episodes/08-development-setup.md:148]: It's still there, right?

File: _episodes/08-development-setup.md [hunk 15]
```diff
@@ -1,138 +1,461 @@
...
+Once our development is finished, we can initiate a ``pull request``.
+For a full description of the GitHub workflow, please see ESMValTool documentation on
+[GitHub Workflow](https://docs.esmvaltool.org/en/latest/community/repository.html#github-workflow).
+
+The pull request will be tested, discussed and merged. This is called "**review process**".
+The process will take some effort and time to learn.
+However, a few “tools” i.e. command lines gets you a long way,
+and we’ll cover those essentials in the next sections.
+
+**Tips**: we encourage you to keep the pull requests small.
```
Comment (Peter9192) [inline _episodes/08-development-setup.md:200]:
```suggestion
**Tip**: we encourage you to keep the pull requests small.
```

File: _episodes/08-development-setup.md [hunk 16]
```diff
@@ -1,138 +1,461 @@
...
+
+We saw 'warming stripes' in lesson
+[Writing your own recipe]({{ page.root }}{% link _episodes/05-preprocessor.md %}).
+Imagine the following task: you want to contribute warming stripes recipe and diagnostics
+to ESMValTool. You have to add the diagnostics
+[warming_stripes.py](../files/warming_stripes.py) and the recipe
+[recipe_warming_stripes.yml](../files/recipe_warming_stripes.yml)
+to their locations in ESMValTool directory.
+After these changes, you should also check if everthing works fine.
+This is where we take advantage of the tools that are introduced later.
```
Comment (Peter9192) [inline _episodes/08-development-setup.md:213]: Just a development tip from my side: I'm using the 'rewrap' extension in vscode. This let's you auto-wrap lines with `alt+Q` so you don't exceed the line length. I find that super useful for these kind of text blocks.

File: _episodes/08-development-setup.md [hunk 17]
```diff
@@ -1,138 +1,461 @@
...
+
+Then, we create a new ``branch`` locally and start developing new codes.
+Once our development is finished, we can initiate a ``pull request``.
+For a full description of the GitHub workflow, please see ESMValTool documentation on
+[GitHub Workflow](https://docs.esmvaltool.org/en/latest/community/repository.html#github-workflow).
+
+The pull request will be tested, discussed and merged. This is called "**review process**".
+The process will take some effort and time to learn.
+However, a few “tools” i.e. command lines gets you a long way,
+and we’ll cover those essentials in the next sections.
```
Comment (Peter9192) [inline _episodes/08-development-setup.md:198]: This sentence is not grammatically correct, can you rephrase it? Suggestion: 
```suggestion
However, a few (command line) tools can get you a long way,
and we’ll cover those essentials in the next sections.
```

File: _episodes/08-development-setup.md [hunk 18]
```diff
@@ -1,138 +1,461 @@
...
+several tools that check our code against those standards like:
 
-The next step is to check that the installation works properly.
-To do this, run the tool with:
+- flake8 for checking against the PEP8 style guide
+- yapf to ensure consistent formatting for the whole project
+- isort to consistently sort the import statements
+- yamllint to ensure there are no syntax errors in our recipes and config files
+- lintr for diagnostic scripts written in R
+- codespell to check grammar
```
Comment (Peter9192) [inline _episodes/08-development-setup.md:227]: Can we make those bold or link to their docs?

```suggestion
- [flake8](https://flake8.pycqa.org/en/latest/#) for checking against the PEP8 style guide
- [yapf](https://pypi.org/project/yapf/) to ensure consistent formatting for the whole project
- [isort](https://pypi.org/project/isort/) to consistently sort the import statements
- [yamllint](https://yamllint.readthedocs.io/en/stable/#) to ensure there are no syntax errors in our recipes and config files
- [lintr](https://github.com/jimhester/lintr) for diagnostic scripts written in R
- [codespell](https://pypi.org/project/codespell/) to check grammar
```

File: _episodes/08-development-setup.md [hunk 19]
```diff
@@ -1,138 +1,461 @@
...
+>> 2. ``flake8``: the error message is about an unused local variable ``nx``.
+>> We should check our codes regarding the usage of ``nx``.
+>> For now, let's assume that it is added by mistake and remove it.
+> {: .solution}
+{: .challenge}
+
+### Run unit tests
+
+Previous section introduced some tools to check code style and quality.
+What it hasen’t done is show us how to tell whether our code is getting the right answer.
```
Comment (Peter9192) [inline _episodes/08-development-setup.md:301]:
```suggestion
What it hasn’t done is show us how to tell whether our code is getting the right answer.
```

File: _episodes/08-development-setup.md [hunk 20]
```diff
@@ -1,138 +1,461 @@
...
+>> as ``warming_stripes.py``:
+>>
+>> ~~~yml
+>>    scripts:
+>>      warming_stripes_script:
+>>        script: warming_stripes.py
+>> ~~~
+>>
+>> For details, see lesson
+>> [Writing your own diagnostic script].
```
Comment (Peter9192) [inline _episodes/08-development-setup.md:364]: Maybe add a dead link for now?
```suggestion
>> [Writing your own diagnostic script](#).
```

File: _episodes/08-development-setup.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (Peter9192) [reply _episodes/08-development-setup.md:372]:
```suggestion
The ESMValTool documentation is available on
```
I think words like documentation and codes are usually 'mass nouns' or 'uncountable nouns', so I would avoid using plural forms like 'codes' or 'documentations'. Just like 'water'.

File: _episodes/08-development-setup.md [hunk 21]
```diff
@@ -1,138 +1,461 @@
...
+### Build documentation
+
+When we add or update a code, we also update its corresponding documentation.
+Documentations are available on
+[docs.esmvaltool.org](https://docs.esmvaltool.org/en/latest/index.html).
+The source files are located in ``ESMValTool/doc/sphinx/source/``.
 
-For a more complete installation verification, run the automated tests and
-confirm that no errors are reported:
+To build documentations locally, first we make sure that the working directory is ``ESMValTool``
```
Comment (Peter9192) [inline _episodes/08-development-setup.md:376]:
```suggestion
To build documentation locally, first we make sure that the working directory is ``ESMValTool``
```

File: _episodes/08-development-setup.md [hunk 22]
```diff
@@ -1,138 +1,461 @@
...
+and our local branch is checked out. Then, we run:
 
 ~~~bash
-python setup.py test --installation
+python setup.py build_sphinx -Ea
 ~~~
 
-> ## ESMValCore
+Similar to codes, documentations should be well written and adhere to standards.
```
Comment (Peter9192) [inline _episodes/08-development-setup.md:383]:
```suggestion
Similar to code, documentation should be well written and adhere to standards.
```

File: _episodes/08-development-setup.md [hunk 23]
```diff
@@ -1,138 +1,461 @@
...
+If the documentation is built properly, the previous command prints a message to the console:
+
+~~~bash
+build succeeded.
+
+The HTML pages are in doc/sphinx/build/html.
+~~~
```
Comment (Peter9192) [inline _episodes/08-development-setup.md:390]:
```suggestion
~~~
build succeeded.

The HTML pages are in doc/sphinx/build/html.
~~~
{: .output}
```

File: _episodes/08-development-setup.md [hunk 24]
```diff
@@ -1,138 +1,461 @@
...
+>>
+>> Then, we build and preview the documentation page:
+>>
+>> ~~~bash
+>> python setup.py build_sphinx -Ea
+>> xdg-open doc/sphinx/build/html/recipes/recipe_warming_stripes.html
+>> ~~~
+>>
+> {: .solution}
+{: .challenge}
```
Comment (Peter9192) [inline _episodes/08-development-setup.md:455]: A closing remark would be nice. Something like

```suggestion
{: .challenge}

Congratulations! You are now ready to make a pull request.
```

**Review by Peter9192 — CHANGES_REQUESTED** [review]
Hey Sarah, thanks for making these changes! I have a couple more small suggestions, mostly typos. Let me know when you're ready, then I'll approve the PR.

File: _episodes/08-development-setup.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply _episodes/08-development-setup.md]: oops, my bad, it is removed now.

**Comment by SarahAlidoost** [comment]
> Hey Sarah, thanks for making these changes! I have a couple more small suggestions, mostly typos. Let me know when you're ready, then I'll approve the PR.

thank you, your comments are addressed.

File: _episodes/08-development-setup.md (hunk 16 above)
Comment (SarahAlidoost) [reply _episodes/08-development-setup.md:211]: I just tried this extension, very cool, thank you for the suggestion.

**Review by Peter9192 — APPROVED** [review]

**Comment by SarahAlidoost** [comment]
Hi @Peter9192 thank you again for reviewing and comments. As we discussed, the failing github action will be fixed in a separate PR. It is mentioned in this issue #191.
